README.md
[![npm](https://img.shields.io/npm/v/vcms.svg)](https://www.npmjs.com/package/vcms)

<h2 align="center"><img src="https://raw.githubusercontent.com/vdegenne/node-vcms/master/logo.png" width="110"><br></h2>
<p  align="center"><strong>A tiny node cms (express, objection, express-session/redis, and more)</strong></p>

## Installation

```bash
yarn add vcms
# or
npm i vcms
```

## Starting a project

create `app.js` (or whichever name you like) :

```javascript
const {start} = require('vcms');
start();
```

That is the minimum code possible for an application using `vcms`.  \
You can run this dummy app using : `node app.js`.
This will just start a server on default port `8000` with nothing but a `ping` route :

```bash
curl localhost:8000/ping
```

If this request returns `"pong"` then it means the server has started successfully.

## Adding some routers

So far our application is boring, does nothing but just "ping/pong".  \
The next step is to add some routes to our `vcms` application.  \
`vcms` organises packs of routes in files also called "routers".  \
To demonstrate this type of organisational structure, let's create `greetings.router.js` in the same directory where we created `app.js` and add this content :

```javascript
const {Router} = require('vcms');

const router = Router();

router.get('/hello', async (req, res) => {
  res.send('hello world')
});

router.get('/bye', async (req, res) => {
  res.send('bye world')
});

module.exports = router;
```

Now we have to tell our application to use this router (which provides 2 routes `/hello` and `/bye`).

Let's create `startupconfig.js` , and write the following :

```javascript
module.exports = (config) => {

  config.routers = {
    '/greetings': require('./greetings.router');
  }

  return config;
}
```

It is important that this file is called `startupconfig.js` here (or `startupconfig.ts` for typescript) because the framework will try to find this file in order to init the application.  \
There is just few requirements, the script needs to return a function. This function has one argument `config`, and finally the function needs to return this `config` object or an object satisfying the `VcmsOptions` interface.  \
Here is an example using ESNext version to make it more clear :

```typescript
import {StartupFunction, VcmsOptions} from 'vcms';
import {greetingsRouter} from './routers/greetings.router';

const startup: StartupFunction = async (config: VcmsOptions) => {

  config.database = true;

  /**
   * config.node_env contains the current NODE_ENV if it was set
   * or is equal to 'prod' as a default
   */
  switch (config.node_env) {
    case 'dev':
      config.db_host = 'localhost:5432'; // shorthand
      // config.db_host will be 'localhost'
      // config.db_port will be 5432
      break;

    case 'prod':
      config.db_host = '1.2.3.4';
      config.db_port = 5433;
      break;
  }

  // routers
  config.routers = {
    '/greetings': greetingsRouter
  }

  // http2 example
  config.http2 = true;
  config.http2_cert = process.cwd() + '/server.crt';
  config.http2_key = process.cwd() + '/server.key';


  return config;
}

export default startup;
```

When we restart our application the routes `/greetings/hello` and `/greetings/bye` should be accessibles.  \
Think about the `startupconfig` script as a middleware we can use to rewrite most of application internal configurations.


*The `startupconfig.js` file is the dynamic way to configure the application. Proceed reading the next chapter to see how to configure the application statically.*



## Static Configuration

One particularity of `vcms` is that it has a default state and this state can be customized almost entirely.  \
In the previous section we saw how to configure the application dynamically (say the behavior defined with some code). There is also a static way and there are three ways of modifying the state statically :

* using a `.vcms.yml` configuration file.
* [dynamically through the `startupconfig.js` file.]
* using environment variables.
* using command-line arguments.

The precedence is performed in the order of the list above. For instance the command-line `--port` argument will override `port` property in the configuration file, and a configuration overridden in `startupconfig.js` will override properties in the `.vcms.yml` file.

### Using command-line arguments

```bash
# Command-Line Arguments
node app.js --port/-p <number> \
            --static <string>  \
            --local-hostname/-h <string> \
            --database/-d [ --db-type <string> \
                            --db-host <string> \
                            --db-port <number> \
                            --db-name <string> \
                            --db-user <string>
                            --db-password <string> ] \
            --session/-s [ --redis-host <string> ] \
            --session-cookie-domain <string> \
            --http2 [ --http2-key <string> \
                      --http2-cert <string> ]
```


All properties are optionals. Between brackets are the properties that are specifics to the property before hand, for instance `--db-type` is only considered by the application if `--database` has been set.

## Configuration options

| Name                  | Type      | Possible values       | Default     | Description                                                                                                                                                                       |
|-----------------------|-----------|-----------------------|-------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| Node Environment      | `string`  | `test`, `dev`, `prod` | `prod`      | Only available as an environmental variable. It is used to load environment-specific options from the configuration file (see *"Configuration File"* section).                    |
| Port                  | `number`  | *                     | 8000        | Listening port of the application.                                                                                                                                                |
| Public Directory      | `string`  | *                     | `public`    | Directory containing static files to serve.                                                                                                                                       |
| Local Hostname        | `string`  | *                     | `localhost` | Local hostname of the local application. It's recommended to use a hostname when using inter-domain cookies or proxy so informations can be shared between urls and applications. |
| Database              | `boolean` | `false`/`true`        | `false`     | Specify if the database support should be enabled or not.                                                                                                                         |
| Session               | `boolean` | `false`/`true`        | `false`     | Specify if the session support should be enabled or not.                                                                                                                          |
| Session Cookie Domain | `string`  | *                     | `localhost` | If the session support is enabled, the main session cookie will use this value as its accessibility scope.                                                                        |

You also have :
- **DB_TYPE**, **DB_HOST**, **DB_PORT**, **DB_NAME**, **DB_USER**, **DB_PASSWORD** ...when the database support is activated
- **REDIS_HOST** ...when the session support is activated


## Configuration file

We can create a file called `.vcms.yml` (yaml file) at the root of our project.

```yaml
port: 3000
local-hostname: hello.app.local

database: true
db-host: localhost:5432

session: false

prod:
  port: 8080
  db-host: 1.2.3.4:5433
  public-directory: public/build
```


## Environmental Variables Equivalent

```bash
# Environmental Variables Version
NODE_ENV
PORT
PUBLIC_DIRECTORY
LOCAL_HOSTNAME
DATABASE
SESSION
SESSION_COOKIE_DOMAIN
HTTP2_REQUIRED
HTTP2_KEY
HTTP2_CERT
```

## notes

- If you want to know more about this framework, please contact me.
- You can propose some PRs on the related github if you want to contribute to this mini project. I'll be more than willing.


CHANGELOG.md
# ChangeLog

## [08-11-2018] v1.9.1

* setDisplay logger for config module

## [05-06-2018] v1.8.5

* Add the `atLeast` option in `validateBody`, `validateParams` helpers.

## [04-06-2018] v1.8.3

* Add the `config` helper property to the `Structure` interface.
* Add `Role` and `User` helper models to the framework.
* Add the `resetDatabase` and access-as helper testing functions.

## [03-06-2018] v1.8.0

* Add the routers `util.ts` file first version.

## [26-05-2018] v1.7.8

* restructuration of `config.ts`.
* Providing the `StartupFunction` type to use with the `startupconfig.js` file to configure the application dynamically.
* `statics` instead of `publics` and add customizable `static` option.
* minor fixes

## [25-05-2018] v1.6.0

* Enhance `publics` configuration capabilities (RegExp support).

## [21-05-2018] v1.5.3

* Structurify the server (see `server.ts` and `Structure` interface). Which eases the testing.
* Add `destroyStructure` function to gracefully close the connections and nullify the structure.

## [21-05-2018] v1.4.0

* Improve the public directories support.
* Add compression middleware.
* Tidy `config.ts` a bit.

## [21-05-2018] v1.3.0

* Add http2 support (`http2`, `http2-key`, `http2-cert`).
* Fix some minor bugs.
* Update `README.md`.


## [18-05-2018] v1.2.0

* dissociate the initialisation of the session in the `getApp` function because it was not modular-proof and was clogging the tests.
* update `README.md`

## [17-05-2018] v1.1.4

* fixes the asynchronous initSessionFunction Middleware.

## [17-05-2018] v1.1.2

* update readme

## [16-05-2018] v1.1.0

* adding startup configuration script support.
* adding `util` test module in npm libraries.

## [15-05-2018] v1.0.6

* logging errors if any happens at start time.

## [15-05-2018] v1.0.1

* update the readme


## [15-05-2018] v1.0.0

* remove `node-env` support in configuration file (as it was prone to design weakness).
* add the `public-directory` option to the configuration file and command line argument (`--public-directory`).
* and some various fixes.
* add this changelog


src/config.ts
import * as commandLineArgs from 'command-line-args';
import { RequestHandler, Router } from 'express';
import { RequestHandlerParams } from 'express-serve-static-core';
import { existsSync, readFileSync } from 'fs';
import { safeLoad } from 'js-yaml';

import commandArgs from './args';
import { Logger } from './logging';


const logger = new Logger('config');

export const setLoggerDisplay = (display: boolean) => logger.setdisplay(display);

export type StartupFunction = (config: VcmsOptions) =>
  Promise<VcmsOptions> | VcmsOptions


/* defaults */
export const defaultOptions: VcmsOptions = {
  node_env: 'prod',
  port: 8000,
  local_hostname: 'localhost',

  http2_required: false,
  http2_key: './server.key',
  http2_cert: './server.crt',

  database_required: false,
  db_host: 'localhost:5432',
  db_type: 'pg',

  session_required: false,
  redis_host: 'localhost:6379'
};



export async function getConfig(
  startupConfigScriptPath?: string,
  configFilepath?: string): Promise<VcmsOptions> {
  let config: VcmsOptions = Object.assign({}, defaultOptions);

  /*===========
   = NODE_ENV =
   ===========*/
  /* from process.env */
  if (process.env.NODE_ENV &&
    ['test', 'dev', 'prod'].includes(process.env.NODE_ENV)) {
    config.node_env = process.env.NODE_ENV
  }


  /* STARTUP CONFIGURATION SCRIPT FILE */
  if (startupConfigScriptPath === undefined) {  // resolving to default
    const possiblePaths = [
      process.cwd(),
      process.cwd() + '/build',
      process.cwd() + '/lib',
      process.cwd() + '/test'
    ];

    const founds =
      possiblePaths.filter(p => existsSync(p + '/startupconfig.js'));

    if (founds.length) {
      startupConfigScriptPath = founds[0] + '/startupconfig.js';
    }
  }

  let startupconfig: VcmsOptions = undefined;
  if (startupConfigScriptPath) {
    if (existsSync(startupConfigScriptPath)) {
      logger.log(`Startup configuration script file resolved to "${
        startupConfigScriptPath}".`);

      const startupfunction = require(startupConfigScriptPath).default;

      if (typeof startupfunction !== 'function')
        throwError(
          'the startup script needs to export a function as the default');

      startupconfig = await startupfunction({ node_env: config.node_env });
      // node_env can be overridden
      if (startupconfig.node_env) {
        config.node_env = startupconfig.node_env;
      }

    } else {
      throwError(`Startup configuration script "${
        startupConfigScriptPath}" couldn't be found`);
    }
  } else {
    logger.info('!! No startup configuration script. Is it expected ?');
  }

  /* CONFIGURATION FILE */
  let userDefinedConfigFilepath = configFilepath ? true : false;
  if (configFilepath === undefined) {
    // resolve to default
    configFilepath = process.cwd() + '/.vcms.yml';
    userDefinedConfigFilepath = false;
  }
  // the startup config script can specify the configuration filepath
  if (startupconfig && startupconfig.configFilepath) {
    configFilepath = startupconfig.configFilepath;
    userDefinedConfigFilepath = true;
  }

  let configFromFile: ConfigFileOptions = undefined;
  if (configFilepath) {
    if (existsSync(configFilepath)) {
      logger.log(`Configuration file resolved to "${configFilepath}".`);

      configFromFile =
        <ConfigFileOptions>safeLoad(readFileSync(configFilepath).toString());
    } else {
      const errMsg =
        `Configuration file "${configFilepath}" couldn't be found.`;
      logger.error(errMsg);
      if (userDefinedConfigFilepath) {
        const error = new Error(errMsg);
        error.name = 'config';
        throw error;
      }
    }
  }


  /* check for command-line options */
  const configFromCommandLine: CommandLineOptions =
    commandLineArgs(commandArgs);


  /*===========
   =   PORT   =
   ===========*/
  // 1. file
  loadOption(
    'port',
    { file: { src: configFromFile, name: 'port', type: 'integer' } },
    config);
  // 2. startupconfig
  transferProperty('port', startupconfig, config);
  // 3. environment and command-line
  loadOption(
    'port',
    {
      env: { src: process.env, name: 'PORT', type: 'integer' },
      cmdline: { src: configFromCommandLine, name: 'port' }
    },
    config);

  /*=====================
   =   LOCAL_HOSTNAME   =
   =====================*/
  // 1. file
  loadOption(
    'local_hostname',
    { file: { src: configFromFile, name: 'local-hostname' } },
    config);
  // 2. startupconfig
  transferProperty('local_hostname', startupconfig, config);
  // 3. environment and command-line
  loadOption(
    'local_hostname',
    {
      file: { src: configFromFile, name: 'local-hostname' },
      env: { src: process.env, name: 'LOCAL_HOSTNAME' },
      cmdline: { src: configFromCommandLine, name: 'local-hostname' }
    },
    config);


  /*====================
   = HTTP2             =
   ====================*/
  // 1. file
  loadOption(
    'http2_required', { file: { src: configFromFile, name: 'http2' } }, config);
  // 2. startupconfig
  transferProperty('http2_required', startupconfig, config);
  // 3. environment and command-line
  loadOption(
    'http2_required',
    {
      env: { src: process.env, name: 'HTTP2_REQUIRED', type: 'boolean' },
      cmdline: { src: configFromCommandLine, name: 'http2' }
    },
    config);

  /*====================
   = HTTP2 CERT        =
   ====================*/
  // 1. file
  loadOption(
    'http2_cert', { file: { src: configFromFile, name: 'http2-cert' } }, config);
  // 2. startupconfig
  transferProperty('http2_cert', startupconfig, config);
  // 3. environment and command-line
  loadOption(
    'http2_cert',
    {
      env: { src: process.env, name: 'HTTP2_CERT' },
      cmdline: { src: configFromCommandLine, name: 'http2-cert' }
    },
    config);



  /*====================
   = HTTP2 KEY         =
   ====================*/
  // 1. file
  loadOption(
    'http2_key', { file: { src: configFromFile, name: 'http2-key' } }, config);
  // 2. startupconfig
  transferProperty('http2_key', startupconfig, config);
  // 3. environment and command-line
  loadOption(
    'http2_key',
    {
      env: { src: process.env, name: 'HTTP2_KEY' },
      cmdline: { src: configFromCommandLine, name: 'http2-key' }
    },
    config);



  /*====================
   = DATABASE_REQUIRED =
   ====================*/
  // 1. file
  loadOption(
    'database_required',
    { file: { src: configFromFile, name: 'database' } },
    config);
  // 2. startupconfig
  transferProperty('database_required', startupconfig, config);
  // 3. environment and command-line
  loadOption(
    'database_required',
    {
      env: { src: process.env, name: 'DATABASE_REQUIRED', type: 'boolean' },
      cmdline: { src: configFromCommandLine, name: 'enable-database' }
    },
    config);

  // in case the database is required
  if (config.database_required) {
    /*===========
     = DB_TYPE  =
     ===========*/
    // 1. file
    loadOption(
      'db_type', { file: { src: configFromFile, name: 'db-type' } }, config);
    // 2. startupconfig
    transferProperty('db_type', startupconfig, config);
    // 3. environment and command-line
    loadOption(
      'db_type',
      {
        env: { src: process.env, name: 'DB_TYPE' }
        // to implement command-line
      },
      config);

    /*===========
     = DB_HOST  =
     ===========*/
    // 1. file
    loadOption(
      'db_host', { file: { src: configFromFile, name: 'db-host' } }, config);
    // 2. startupconfig
    transferProperty('db_host', startupconfig, config);
    // 3. environment and command-line
    loadOption(
      'db_host',
      {
        env: { src: process.env, name: 'DB_HOST' }
        // to implement command-line
      },
      config);

    // format DB_HOST in case it contains the port
    if (config.db_host.indexOf(':') > -1) {
      const dbHostParts = config.db_host.split(':');
      config.db_host = dbHostParts[0];
      config.db_port = parseInt(dbHostParts[1]);
    }

    /*===========
     = DB_PORT  =
     ===========*/
    // 1. file
    loadOption(
      'db_port', { file: { src: configFromFile, name: 'db-port' } }, config);
    // 2. startupconfig
    transferProperty('db_port', startupconfig, config);
    // 3. environment and command-line
    loadOption(
      'db_port',
      {
        env: { src: process.env, name: 'DB_PORT' },
        cmdline: { src: configFromCommandLine, name: 'db-port' }
      },
      config);

    // if no DB_PORT was found, resolve based on the type
    if (!config.db_port) {
      switch (config.db_type) {
        case 'pg':
          config.db_port = 5432;
      }
    }

    /*===========
     = DB_NAME  =
     ===========*/
    // 1. file
    loadOption(
      'db_name', { file: { src: configFromFile, name: 'db-name' } }, config);
    // 2. startupconfig
    transferProperty('db_name', startupconfig, config);
    // 3. environment and command-line
    loadOption(
      'db_name',
      {
        env: { src: process.env, name: 'DB_NAME' },
        cmdline: { src: configFromCommandLine, name: 'db-name' }
      },
      config);

    /*===========
     = DB_USER  =
     ===========*/
    // 1. file
    loadOption(
      'db_user', { file: { src: configFromFile, name: 'db-user' } }, config);
    // 2. startupconfig
    transferProperty('db_user', startupconfig, config);
    // 3. environment and command-line
    loadOption(
      'db_user',
      {
        env: { src: process.env, name: 'DB_USER' },
        cmdline: { src: configFromCommandLine, name: 'db-user' }
      },
      config);

    /*===============
     = DB_PASSWORD  =
     ===============*/
    // 1. file
    loadOption(
      'db_password',
      { file: { src: configFromFile, name: 'db-password' } },
      config);
    // 2. startupconfig
    transferProperty('db_password', startupconfig, config);
    // 3. environment and command-line
    loadOption(
      'db_password',
      {
        env: { src: process.env, name: 'DB_PASSWORD' },
        // to implement command-line
      },
      config);
  }

  /*===================
   = SESSION_REQUIRED =
   ===================*/
  // 1. file
  loadOption(
    'session_required',
    { file: { src: configFromFile, name: 'session' } },
    config);
  // 2. startupconfig
  transferProperty('session_required', startupconfig, config);
  // 3. environment and command-line
  loadOption(
    'session_required',
    {
      env: { src: process.env, name: 'SESSION_REQUIRED', type: 'boolean' },
      cmdline: { src: configFromCommandLine, name: 'enable-session' }
    },
    config);

  // if session is required, specific options
  if (config.session_required) {
    /*=============
     = REDIS_HOST =
     =============*/
    // 1. file
    loadOption(
      'redis_host',
      { file: { src: configFromFile, name: 'redis-host' } },
      config);
    // 2. startupconfig
    transferProperty('redis_host', startupconfig, config);
    // 3. environment and command-line
    loadOption(
      'redis_host',
      {
        env: { src: process.env, name: 'REDIS_HOST' },
        cmdline: { src: configFromCommandLine, name: 'redis-host' }
      },
      config);


    /*============================
     =   SESSION_COOKIE_DOMAIN   =
     ============================*/
    // 1. file
    loadOption(
      'session_cookie_domain',
      { file: { src: configFromFile, name: 'session-cookie-domain' } },
      config);
    // 2. startupconfig
    transferProperty('session_cookie_domain', startupconfig, config);
    // 3. environment and command-line
    loadOption(
      'session_cookie_domain',
      {
        env: { src: process.env, name: 'SESSION_COOKIE_DOMAIN' },
        cmdline: { src: configFromCommandLine, name: 'session-cookie-domain' }
      },
      config);
  }


  /*=======================
   =   static             =
   =======================*/
  // 1. file
  loadOption('static', { file: { src: configFromFile, name: 'static' } }, config);
  // 2. startupconfig
  transferProperty('static', startupconfig, config);
  // 3. environment and command-line
  loadOption(
    'static',
    {
      env: { src: process.env, name: 'STATIC' },
      cmdline: { src: configFromCommandLine, name: 'static' }
    },
    config);


  /*=======================
   =   statics            =
   =======================*/
  // 1. file
  loadOption('statics', { file: { src: configFromFile, name: 'statics' } }, config);
  // 2. startupconfig
  transferProperty('statics', startupconfig, config);
  // we should convert the regexp routes
  if (config.statics) {
    for (const s of config.statics) {
      s.route = <string>s.route;
      if (typeof s.route !== 'object' && s.route !== '/' &&
        s.route.startsWith('/\\/') && s.route.endsWith('/')) {
        s.route = new RegExp(s.route.replace(/^\/|\/$/g, ''));
      }
    }
  }


  // add startup configurations to the main configuration object
  config = { ...config, ...startupconfig };

  // we finally return the config
  return <VcmsOptions>config;
}


export type Static = {
  route: string | RegExp,
  serve: string
}


export interface VcmsOptions {
  node_env: string;
  port: number;
  local_hostname: string;

  http2_required: boolean;
  http2_key?: string;
  http2_cert?: string;


  database_required: boolean;
  db_type?: string;
  db_host?: string;
  db_port?: number;
  db_name?: string;
  db_user?: string;
  db_password?: string;

  session_required: boolean;
  redis_host?: string;
  session_cookie_domain?: string;

  configFilepath?: string;
  routers?: { [base: string]: Router | RequestHandler };
  initSessionFunction?: (session: Express.Session) => void;
  static?: string;
  statics?: Static[];
  middlewares?: RequestHandler[];
}


export interface ConfigFileOptions extends ConfigFileOptionsBase {
  prod: ConfigFileOptionsBase;
  dev: ConfigFileOptionsBase;
  test: ConfigFileOptionsBase;
}


export interface ConfigFileOptionsBase {
  port?: number;
  'local-hostname'?: string;

  http2?: boolean;
  'http2-key'?: string;
  'http2-cert'?: string;

  database?: boolean;
  'db-type'?: string;
  'db-host'?: string;
  'db-port'?: string;
  'db-name'?: string;
  'db-user'?: string;
  'db-password'?: string;

  session?: boolean;
  'redis-host'?: string;
  'session-cookie-domain'?: string;

  static?: string;
  'statics'?: Static[];
}

export interface CommandLineOptions {
  port?: number;
  'local-hostname'?: string;

  http2?: boolean;
  'http2-key'?: string;
  'http2-cert'?: string;

  'enable-database'?: boolean;
  'db-port'?: number;
  'db-name'?: string;
  'db-user'?: string;
  'redis-host'?: string;
  'session-cookie-domain'?: string;

  static?: string;
}

interface OptionSpecifier {
  name: string;
  src?: ConfigFileOptions | CommandLineOptions;
  type?: string
}

function loadOption(
  optionName: string,
  from: {
    file?: OptionSpecifier,
    env?: OptionSpecifier,
    cmdline?: OptionSpecifier
  },
  config: VcmsOptions) {
  /* from file */
  if (from.file && from.file.src) {
    const src = from.file.src;
    const name = from.file.name;
    if (src[config.node_env] && src[config.node_env][name]) {
      config[optionName] = src[config.node_env][name];
    } else if (src[name]) {
      config[optionName] = src[name];
    }
  }

  /* from process.env */
  if (from.env && process.env[from.env.name]) {
    config[optionName] = from.env.src[from.env.name];
    if (from.env.type === 'boolean') {
      config[optionName] =
        config[optionName].toLowerCase() === 'true' ? true : false;
    }
    if (from.env.type === 'integer') {
      config[optionName] = parseInt(config[optionName]);
    }

    // DOCKER FOR DB_HOST
    if (optionName === 'DB_HOST' && config[optionName] === 'DOCKER_HOST') {
      if (!from.env.src['DOCKER_HOST']) {
        throw new Error('DOCKER_HOST is not defined.');
      }
      config[optionName] = from.env.src['DOCKER_HOST'];
    }

    // DOCKER FRO REDIS_HOST
    if (optionName === 'REDIS_HOST' && config[optionName] === 'DOCKER_HOST') {
      if (!from.env.src['DOCKER_HOST']) {
        throw new Error('DOCKER_HOST is not defined.');
      }
      config[optionName] = from.env.src['DOCKER_HOST'];
    }
  }

  /* from command line */
  if (from.cmdline && from.cmdline.src && from.cmdline.src[from.cmdline.name]) {
    config[optionName] = from.cmdline.src[from.cmdline.name];
  }
}


function transferProperty(
  propName: string, from: VcmsOptions, to: VcmsOptions) {
  if (!from || !(propName in from)) {
    return;
  }

  to[propName] = from[propName];
  delete from[propName];
}


function throwError(message: string) {
  logger.error(message);
  const error = new Error(message);
  error.name = 'config';
  throw error;
}
